src/treebuilders/multiply.cpp

```diff
@@ -290,6 +292,46 @@ template <int D> double dot(FunctionTree<D> &bra, FunctionTree<D> &ket) {
     return result;
 }
 
+/** @brief abs-dot product of two MW function representations
+ *
+ * @param[in] bra Bra side input function
+ * @param[in] ket Ket side input function
+ *
+ * If exact=false: does not at any time read the coefficients individually.
+ * The product is done for the end nodes of the bra multiplied by the nodes from the
+ * ket with either the same idx, or using a lower scale and assuming uniform
+ * distribution within the node. If the product is zero, the functions are disjoints.
+ */
+template <int D> double node_norm_dot(FunctionTree<D> &bra, FunctionTree<D> &ket, bool exact) {
+
+    double result = 0.0;
+    int ncoef = bra.getKp1_d() * bra.getTDim();
+    double valA[ncoef];
+    double valB[ncoef];
+    int nNodes = bra.getNEndNodes();
+
+    if (exact and nNodes != ket.getNEndNodes()) { MSG_ABORT("Trees must have same grid"); }
+    for (int n = 0; n < nNodes; n++) {
+        FunctionNode<D> &node = bra.getEndFuncNode(n);
+        const NodeIndex<D> idx = node.getNodeIndex();
+        FunctionNode<D> *mwNode = (FunctionNode<D> *)ket.findNode(idx);
```

Comment:
please use `static_cast`

Reply:
> Do I understand correctly that even `exact=true` is asymmetric in the bra and ket? It's always the bra that defines the grid, and no nodes are generated in the process?

If `exact=true`, it is required that the grid are already defined equal by the calling routine. `if (exact and nNodes != ket.getNEndNodes()) { MSG_ABORT("Trees must have same grid"); }`

Reply:
> please use `static_cast`

done

Reply:
I see. It's not a fool proof check, though, as different trees can have the same number of nodes. You could rather check for `nullptr` in the `findNode()` return value

Reply:
done